src/Http/Environment.php

```
@@ -0,0 +1,904 @@
+     * @return void
+     */
+    protected function detectAuto(bool $insecure = false): void
+    {
+        // proxy server setup
+        if ($this->isBehindProxy === true && $insecure === true) {
+            $this->host  = $this->detectForwardedHost();
+            $this->https = $this->detectForwardedHttps();
+            $this->port  = $this->detectForwardedPort();
+
+        // local server setup
+        } else {
+            $this->host  = $this->detectHost($insecure);
+            $this->https = $this->detectHttps();
+            $this->port  = $this->detectPort();
+        }
+    }
+
+    /**
+     * @param bool|null $override
+     * @return bool
+     */
+    protected function detectCli(?bool $override = null): bool
+    {
+        if ($override === false) {
+            return false;
+        }
+
+        if ($override === true) {
+            return true;
+        }
+
+        if (defined('STDIN') === true) {
+            return true;
+        }
+
+        $term = getenv('TERM');
+
+        if (substr(PHP_SAPI, 0, 3) === 'cgi' && $term && $term !== 'unknown') {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * @return string|null
+     */
+    protected function detectForwardedHost(): ?string
+    {
+        $host = $this->get('HTTP_X_FORWARDED_HOST');
+
+        if (empty($host) === false) {
+            $parts = $this->detectPortInHost($host);
+
+            $this->portInHost = $parts['port'];
+
+            return $parts['host'];
+        }
+
+        return $this->detectHost(true);
+    }
+
+    /**
+     * @return bool
+     */
+    protected function detectForwardedHttps(): bool
+    {
+        if ($this->detectHttpsOn($this->get('HTTP_X_FORWARDED_SSL')) === true) {
+            return true;
+        }
+
+        if ($this->detectHttpsProtocol($this->get('HTTP_X_FORWARDED_PROTO')) === true) {
+            return true;
+        }
+
+        // fall back to local setup
+        return $this->detectHttps();
```

Not sure about this. If this line is ever reached, the result will most likely be incorrect as the local HTTPS status doesn't tell anything about the actual protocol that the user sees on the outside. I feel like we should rather throw an Exception if this is reached. Or have I missed anything?

---

TBH I'm not experienced enough with proxy setups. If the local setup really does not provide any useful reference in such a case, we can really skip it. Wouldn't it be fine to just assume it's not HTTPS in that case instead of throwing an exception?

---

I think we cannot assume anything really. If the proxy sets the `X-Forwarded-Host` header but none of the headers we can use for port and HTTPS detection, this would IMO either be a misconfiguration in the proxy or a bug in our detection code. So I think we should make it clear that something is wrong.

---

I just checked again and in a non-proxy setup, $_SERVER['HTTPS'] is undefined, if there's no HTTPS setup. IMHO the correct solution would be to simply return false instead of falling back to the local setup. This way, the proxied HTTPS status is considered to be true if it is set correctly, but false if it is undefined.